### PR TITLE
Classic to Core format converter

### PR DESCRIPTION
**Pull request by Scarjit**
This processor plugin converts classic _historian data to umh-core compatible formats.
Basically it unpacks jsons containing timestamps and multiple values, and does some work on the topics.

For more details please also checkout it's documentation :)

### CODE REVIEW

File: docs/processing/classic-to-core-processor.md [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+## Configuration
+
+```yaml
+pipeline:
+  processors:
+    - classic_to_core:
+        timestamp_field: timestamp_ms # Field containing timestamp (default: timestamp_ms)
+        target_data_contract: _raw # Target data contract. If empty, uses input's contract
+        exclude_fields: [] # List of fields to exclude from conversion
+        preserve_meta: true # Whether to preserve original metadata
```
[the remark this answers is not in the document]
Comment (Scarjit) [reply docs/processing/classic-to-core-processor.md]: Not sure, what you mean by this ?

File: docs/processing/classic-to-core-processor.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+The Classic to Core Processor transforms single messages containing multiple values and timestamps into separate Core format messages, each containing a single value. It handles nested tag groups by flattening them into intuitive dot-notation paths and reconstructs topics according to Core conventions.
+
+## When to Use
+
+Use the `classic_to_core` processor when you need to:
+
+- **Process Classic Data in Core**: Convert Classic UMH Historian schema messages to Core format for processing in Core-based systems
+- **Use Core Processors**: Enable the use of processors that rely on the Core data model (like downsampler)
+
+The processor maintains compatibility with the `_historian` schema, ensuring that the output can be consumed by systems expecting Classic format messages.
```
Comment (JeremyTheocharis) [inline docs/processing/classic-to-core-processor.md:16]: are you sure? the following processors expect UMH Core

File: docs/processing/classic-to-core-processor.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+| Parameter              | Type   | Default | Description                                         |
+| ---------------------- | ------ | ------- | --------------------------------------------------- |
+| `target_data_contract` | string | `""`    | Target data contract. If empty, uses input's schema |
+
+## Message Transformation
+
+### Basic Conversion
+
+The processor transforms single messages with multiple values into separate Core format messages:
+
```
Comment (JeremyTheocharis) [inline docs/processing/classic-to-core-processor.md:37]: add that target_data_contract here is "_raw"

File: classic_to_core_plugin/example_config.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        exclude_fields:
+          - _quality
+          - _debug_info
+        max_recursion_depth: 10
+        max_tags_per_message: 1000
+
+output:
+  # Publish to individual Core topics using UNS output
+  uns_output:
+    kafka:
```
Comment (JeremyTheocharis) [inline classic_to_core_plugin/example_config.yaml:42]: probably AI left over

File: classic_to_core_plugin/example_config.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    addresses: ['localhost:9092']
+    topics: ['umh.v1.+.+.+._historian.+']
+    consumer_group: 'historian-to-core-migration'
+    start_from_oldest: true
+
+pipeline:
+  processors:
+    # Convert Historian Data Contract format to Core format
+    - classic_to_core:
+        timestamp_field: timestamp_ms
```
Comment (JeremyTheocharis) [inline classic_to_core_plugin/example_config.yaml:30]: old sample format

**Review by JeremyTheocharis — CHANGES_REQUESTED** [review]

**Review by JeremyTheocharis — CHANGES_REQUESTED** [review]
Here is also an O3 review. Do with it what you wnat, burt there are some good points in there.

maybe also add it to CI so we chave an automated terst here as well

<html><head></head><body><p>Below is a structured, point‑by‑point review of the <strong><code inline="">classic_to_core</code></strong> contribution.<br>
I focus on API design, functional correctness, edge‑case coverage and maintainability, and I cross‑check with the public specifications for <strong>UMH Core</strong> and <strong>UMH Classic / Historian</strong>.</p>
<hr>
<h2>1 &nbsp;Public‑facing contract vs. implementation</h2>

Area | Observation | Impact | Recommendation
-- | -- | -- | --
Config spec | example_config.yaml and docs list options such as timestamp_field, preserve_meta, exclude_fields, max_recursion_depth, max_tags_per_message, yet NewConfigSpec() only exposes target_data_contract. | Users following the docs will get unknown field errors. | Align the two sources of truth: either implement the additional fields (preferred) or drop them from docs/example.
Defaults hard‑coded | Recursion depth 10, tag limit 1000, metadata preservation “always on”, excluded keys meta, metadata, timestamp_ms – all are constants, not configurable. | Can’t tune processor for huge telemetry packets or stricter installations. | Promote the constants to config fields with sane defaults (backward compatible).
Timestamp field | Code assumes hard‑coded timestamp_ms; examples show it is configurable. | Payloads that use ts_ms or similar will be silently rejected. | Add an optional timestamp_field config entry; keep default timestamp_ms.


<hr>
<h2>4 &nbsp;Security &amp; robustness</h2>
<ul>
<li>
<p>No reflection or unsafe code.</p>
</li>
<li>
<p><code inline="">json.Marshal</code> on arbitrary <code inline="">interface{}</code> can blow up if value contains channels / funcs, but Benthos structured payload can’t carry those – acceptable.</p>
</li>
<li>
<p>Recursion guard avoids maliciously deep objects.</p>
</li>
</ul>
<hr>
<h2>5 &nbsp;Missing or weakly‑tested edge cases</h2>
<ol>
<li>
<p><strong>Array payload members</strong> (should reject or split).</p>
</li>
<li>
<p><strong>Very large numeric timestamps</strong> that cannot be represented exactly by <code inline="">float64</code>.</p>
</li>
<li>
<p><strong>Schema starting later in the path</strong> (<code inline="">umh.v1.dept._north.section._historian.context</code>) – only first <code inline="">_</code> is used; need test.</p>
</li>
<li>
<p><strong>UTF‑8 / special characters in tag names</strong>.</p>
</li>
<li>
<p><strong>Duplicate keys after flattening</strong> (clash detection).</p>
</li>
</ol>
<hr>
<h2>6 &nbsp;Recommended next steps</h2>
<ol>
<li>
<p><strong>Finish the configuration surface</strong> – expose all documented knobs and add validation (<code inline="">min&gt;0</code>, etc.).</p>
</li>
<li>
<p><strong>Strict scalar value check</strong> before message creation; add <code inline="">messages_rejected_non_scalar</code> metric.</p>
</li>
<li>
<p><strong>Topic parser hardening</strong> – detect multiple schemas, invalid characters, unknown schemas.</p>
</li>
<li>
<p><strong>Improve meta flattening</strong> – optionally allow nested meta objects to be flattened with prefix, or reject them; stringifying a map into <code inline="">map[...]</code> is not helpful.</p>
</li>
<li>
<p><strong>Docs ⇄ Code sync</strong> – run <code inline="">make docs-lint</code> (if you have such target) in CI to flag drift.</p>
</li>
<li>
<p><strong>Benchmark</strong> – add <code inline="">go test -bench</code> to measure flattening vs. payload size.</p>
</li>
</ol>
<hr>
<h3>TL;DR</h3>
<p><em>Implementation is sound and well‑tested for the common path, but the public configuration and the documented behaviour diverge; non‑scalar values and some corner‑case topics can slip through in ways that violate UMH‑Core rules.</em><br>
Address the config mismatch first, then tighten validation around arrays / unknown value types and duplicate flattened keys. After that, the processor will be production‑ready.</p>
<p>Hope this helps to polish the plugin!</p></body></html>Below is a structured, point‑by‑point review of the **`classic_to_core`** contribution.
I focus on API design, functional correctness, edge‑case coverage and maintainability, and I cross‑check with the public specifications for **UMH Core** and **UMH Classic / Historian**.

---

## 1  Public‑facing contract vs. implementation

| Area                    | Observation                                                                                                                                                                                                             | Impact                                                                     | Recommendation                                                                                                       |
| ----------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------- |
| **Config spec**         | `example_config.yaml` and docs list options such as `timestamp_field`, `preserve_meta`, `exclude_fields`, `max_recursion_depth`, `max_tags_per_message`, yet `NewConfigSpec()` only exposes **`target_data_contract`**. | Users following the docs will get *unknown field* errors.                  | *Align the two sources of truth:* either implement the additional fields (preferred) or drop them from docs/example. |
| **Defaults hard‑coded** | Recursion depth 10, tag limit 1000, metadata preservation “always on”, excluded keys `meta`, `metadata`, `timestamp_ms` – all are constants, not configurable.                                                          | Can’t tune processor for huge telemetry packets or stricter installations. | Promote the constants to config fields with sane defaults (backward compatible).                                     |
| **Timestamp field**     | Code assumes hard‑coded `timestamp_ms`; examples show it is configurable.                                                                                                                                               | Payloads that use `ts_ms` or similar will be silently rejected.            | Add an optional `timestamp_field` config entry; keep default `timestamp_ms`.                                         |

---

## 2  Correctness & edge‑case coverage

### 2.1 Core semantics

*The processor must output “one tag, one message, one topic” and accept only scalar values (number, boolean, string) as UMH‑Core “time‑series / tags”.* ([[docs.umh.app](https://docs.umh.app/usage/unified-namespace/payload-formats)][1])

| Edge case                                                                                  | Current behaviour                                        | Spec‑compliance                                                              |
| ------------------------------------------------------------------------------------------ | -------------------------------------------------------- | ---------------------------------------------------------------------------- |
| **Array values** (`[]interface{}`)                                                         | Treated as plain value and forwarded.                    | ❌  UMH‑Core spec allows *scalar* only; array should raise error or be split. |
| **Object values** deeper than depth limit                                                  | Flattening stops and logs warning.                       | ✔ (limit is reasonable)                                                      |
| **Duplicate flattened keys** (name clash between parent.child and existing `parent.child`) | Last one wins (silent overwrite).                        | ⚠  Consider warning or deterministic suffix.                                 |
| **Field names containing “.” already**                                                     | No special handling; becomes `outer.inner.key.with.dot`. | ✔ (accepted by topic rules)                                                  |
| **Meta / metadata non‑object**                                                             | Skipped silently (tests cover).                          | ✔                                                                            |

### 2.2 Topic parsing

*UMH topic grammar*: `umh.v1.<location_path>._<schema>.<virtual_path?>`

Parsing is correct in the happy path, but:

* It always takes the **first** `_word` as schema. If a location segment itself starts with `_` you get a false positive.
* No check that schema itself is whitelisted (e.g., `_historian`, `_raw`, `_cnc`, …). Optional, but would avoid typos.

### 2.3 Metrics & logging

*Good:* Counters for processed / errored / expanded / truncated.
*Minor:* `messagesDropped` is incremented when tag limit is exceeded **both** inside `processMessage` (truncate) **and** again in `ProcessBatch` (drop), but only the latter drops. This double account can mislead dashboards.

---

## 3  API & code quality

| Item                  | Note                                                                                                                    |
| --------------------- | ----------------------------------------------------------------------------------------------------------------------- |
| **Test suite**        | Very comprehensive: >1 800 LOC, covers most branches. Nice use of Ginkgo tables.                                        |
| **Env gating**        | Tests require `TEST_CLASSIC_TO_CORE=1`; CI job must export it or the suite is skipped.                                  |
| **Docs**              | Tutorial quality – great. But because config/spec mismatch, trust is eroded.                                            |
| **panic in `init()`** | `RegisterBatchProcessor` failure panics. This is Benthos‑idiomatic and fine.                                            |
| **Error messages**    | Precise and user‑oriented. Good.                                                                                        |
| **Performance**       | Minimal allocations, but per‑message deep copy of metadata + JSON re‑marshal is inevitable. No obvious hot‑path issues. |

---

## 4  Security & robustness

* No reflection or unsafe code.
* `json.Marshal` on arbitrary `interface{}` can blow up if value contains channels / funcs, but Benthos structured payload can’t carry those – acceptable.
* Recursion guard avoids maliciously deep objects.

---

## 5  Missing or weakly‑tested edge cases

1. **Array payload members** (should reject or split).
2. **Very large numeric timestamps** that cannot be represented exactly by `float64`.
3. **Schema starting later in the path** (`umh.v1.dept._north.section._historian.context`) – only first `_` is used; need test.
4. **UTF‑8 / special characters in tag names**.
5. **Duplicate keys after flattening** (clash detection).

---

## 6  Recommended next steps

1. **Finish the configuration surface** – expose all documented knobs and add validation (`min>0`, etc.).
2. **Strict scalar value check** before message creation; add `messages_rejected_non_scalar` metric.
3. **Topic parser hardening** – detect multiple schemas, invalid characters, unknown schemas.
4. **Improve meta flattening** – optionally allow nested meta objects to be flattened with prefix, or reject them; stringifying a map into `map[...]` is not helpful.
5. **Docs ⇄ Code sync** – run `make docs-lint` (if you have such target) in CI to flag drift.
6. **Benchmark** – add `go test -bench` to measure flattening vs. payload size.

---

### TL;DR

*Implementation is sound and well‑tested for the common path, but the public configuration and the documented behaviour diverge; non‑scalar values and some corner‑case topics can slip through in ways that violate UMH‑Core rules.*
Address the config mismatch first, then tighten validation around arrays / unknown value types and duplicate flattened keys. After that, the processor will be production‑ready.

Hope this helps to polish the plugin!

[1]: https://docs.umh.app/usage/unified-namespace/payload-formats "Payload Formats | UMH Docs"

File: classic_to_core_plugin/classic_to_core_plugin_test.go [hunk 6]
```diff
@@ -0,0 +1,1988 @@
...
+var _ = Describe("ClassicToCoreProcessor", func() {
+	BeforeEach(func() {
+		testActivated := os.Getenv("TEST_CLASSIC_TO_CORE")
+
+		// Check if environment variables are set
+		if testActivated == "" {
+			Skip("Skipping Classic to Core tests: TEST_CLASSIC_TO_CORE not set")
+			return
+		}
+	})
```
[the remark this answers is not in the document]
Comment (Scarjit) [reply classic_to_core_plugin/classic_to_core_plugin_test.go:41]: This is the same behaviour as other test suites.

**Comment by Scarjit** [comment]
bugbot run

**Review by JeremyTheocharis — APPROVED** [review]
